packages/salesforcedx-apex-replay-debugger/src/core/heapDumpService.ts

```diff
@@ -136,49 +127,123 @@ export class HeapDumpService {
     }
   }
 
-  public updateVariableContainer(
+  public updateContainerOrCache(
+    variableContainer: ApexVariableContainer,
+    variableEntry: HeapDumpExtentValueEntry,
+    extentToRevisit: Map<string, HeapDumpExtentValue>,
+    referenceToRevisit: Map<string, ApexVariableContainer>,
+    variableTypes: Map<string, string>
+  ): void {
+    if (this.isAddress(variableEntry.value.value)) {
+      if (extentToRevisit.has(variableEntry.value.value)) {
+        const extentValue = extentToRevisit.get(variableEntry.value.value);
+        if (extentValue) {
+          this.updateContainerType(
+            variableContainer,
+            variableTypes.get(variableEntry.value.value)
+          );
+          this.updateContainer(
+            variableContainer,
+            extentValue,
+            extentToRevisit,
+            referenceToRevisit,
+            variableTypes
+          );
+        }
+      } else {
+        referenceToRevisit.set(variableEntry.value.value, variableContainer);
+      }
+    } else {
+      this.updateContainerWithExtentValueOrEntry(
+        variableContainer,
+        variableEntry.value.value,
+        undefined,
+        variableTypes.get(variableEntry.keyDisplayValue)
+      );
+    }
+  }
+
+  public updateContainer(
     varContainer: ApexVariableContainer,
-    heapDumpExtentValue: HeadpDumpExtentValue,
-    type?: string
+    heapDumpExtentValue: HeapDumpExtentValue,
+    extentToRevisit: Map<string, HeapDumpExtentValue>,
+    referenceToRevisit: Map<string, ApexVariableContainer>,
+    variableTypes: Map<string, string>
   ): void {
     if (heapDumpExtentValue.value) {
```

Comment:
Whenever we're checking for the existence of an actual variable value we should be using if (typeof heapDumpExtentValue.value !== 'undefined'). The reason for this is that if (heapDumpExtentValue.value) would evaluate to false if the value was a boolean set to false.